### PR TITLE
Extensible Messaging and Presence Protocol - XMPP with NodeJs and Docker

### LINKED ISSUE #4943: Extensible Messaging and Presence Protocol - XMPP with NodeJs and Docker
### Introduction

XMPP - Extensible Messaging and Presence Protocol, originally known as `ejabberd` is an open communication protocol designed for instant messaging presence information and contact list maintenance.  It is used by almost all large messaging systems such as WhatsApp, Facebook, Google-Doc, etc. 

This tutorial will explain the XMPP architecture, how it works, and build a demo server called ejaberb using nodejs and docker commands. 

### Key takeaways
At the end of this tutorial, readers would have knowledge of the following points;
1. What is XMPP and why it exits
2. Architecture and how it works
3. Building XMPP server with nodejs and docker

### Article Quality
This tutorial targets the explanations behind the key messaging presence protocol where XMPP is an instance. This will not only explain its architecture and how it works underline but also build a demo that readers can make use of as part of the portfolio.

### References
N/A

### PR DESCRIPTION
**Pull request by Horlawhumy-dev**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes
This closes #4943

### CODE REVIEW

**Comment by jumamiller** [comment]
@Horlawhumy-dev, ensure that all code snippets are well explained!

**Comment by Horlawhumy-dev** [comment]
I will review it ASAP. Thank you

**Comment by Horlawhumy-dev** [comment]
@jumamiller kindly check it out.

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

**Comment by Horlawhumy-dev** [comment]
@jumamiller sir!!!

File: content/articles/extensible-messaging-and-presence-protocol-with-nodejs-and-docker/index.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+There are three different XML stanzas that XMPP supports.
+
+1. Messages can be sent with XML tag <message/>
+2. The status can also be exchanged like <presence/>
+3. Information query which is the request-response like <iq/> tag.
+
+Finally, the whole of XMPP is decentralized. For instance, we have the domain `myxmpp.com` server between two clients. See the instance below on how the connection works :)
+
+ClientA concects to xmpp server,
+
```
Comment (jumamiller) [inline content/articles/extensible-messaging-and-presence-protocol-with-nodejs-and-docker/index.md:53]: what do you mean with term `concects`,connect?

File: content/articles/extensible-messaging-and-presence-protocol-with-nodejs-and-docker/index.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+```xml
+<message from='a@myxmpp.com'>
+    <body>Hi </body>
+</message>
+```
+
+ClientA terminates the stream
+```xml
+</stream:stream'>
```
Comment (jumamiller) [inline content/articles/extensible-messaging-and-presence-protocol-with-nodejs-and-docker/index.md:89]: check the apostrophe? do we need it?

**Review by jumamiller — CHANGES_REQUESTED** [review]
Excellent work, however your article lacks introduction! I would also love it if you added a table of contents for more straightforward navigation!

**Comment by Horlawhumy-dev** [comment]
Thank you @jumamiller 

I have added some explanations in the introduction section.

**Review by jumamiller — CHANGES_REQUESTED** [review]
@Horlawhumy-dev, could we have full paragraphs? I noticed you have a lot of spacings which makes it hard to follow the article. Reformat them well.

**Comment by Horlawhumy-dev** [comment]
Formatting done.

**Review by jumamiller — APPROVED** [review]
@Horlawhumy-dev, check the changes made during the review t improve your future articles!

**Comment by Horlawhumy-dev** [comment]
I have gone through them. Thank you, sir!

**Comment by hectorkambow** [comment]
@jumamiller  I think you are doing too much during your reviews. 

We need to be mindful of all our pull requests and all our students.
